### PR TITLE
Add Liquid Web DNS Provider

### PR DESCRIPTION
**Pull request by caiges**
I've implemented a DNS solver using [Liquid Web's API](https://cart.liquidweb.com/storm/api/docs/v1/)

### CODE REVIEW

**Comment by ldez** [comment]
Hello, in order for a PR adding a DNS provider to be accepted, you have to:

- [ ] add a description to your PR
- [ ] be able to maintain this provider
- [ ] have a homogeneous design with the other providers
- [ ] add tests ("live" and units)
https://github.com/go-acme/lego/blob/c938de68f262c057542d44bcb67ef961cefcc60e/providers/dns/cloudflare/cloudflare_test.go#L125-L151

```shell
make test
```
- [ ] add the provider to the [readme.md](https://github.com/go-acme/lego#dns-providers)
- [ ] add a [provider descriptor](https://github.com/go-acme/lego/blob/master/providers/dns/cloudflare/cloudflare.toml)
- [ ] generate CLI help and documentation
```bash
make generate-dns
```
- [ ] be able to do: _(and put the output of this command to a comment in your PR)_
```bash
rm -rf .lego

./lego -m your@email.com --dns YOUR_PROVIDER_NAME -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```
Note the wildcard domain is important.
- [ ] pass the linter ([golangci-lint](https://github.com/golangci/golangci-lint#install) must be installed):
```shell
make checks
```

File: providers/dns/liquidweb/liquidweb.toml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+Name = "Liquid Web"
+Description = ''''''
+URL = "https://cart.liquidweb.com/storm/api/docs/v1/"
+Code = "liquidweb"
+Since = "v3.1.0"
+
+Example = ''''''
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.toml]: could you add a command example.

File: providers/dns/liquidweb/liquidweb.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+// NewDefaultConfig returns a default configuration for the DNSProvider
+func NewDefaultConfig() *Config {
+	config := &Config{
+		URL:                env.GetOrDefaultString("LW_URL", ""),
+		Username:           env.GetOrDefaultString("LW_USERNAME", ""),
+		Password:           env.GetOrDefaultString("LW_PASSWORD", ""),
+		Zone:               env.GetOrDefaultString("LW_ZONE", ""),
+		Timeout:            time.Second * time.Duration(env.GetOrDefaultInt("LW_TIMEOUT", 60)),
+		PropagationTimeout: time.Minute * 5,
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]: Could you follow the same pattern as other provider:
https://github.com/go-acme/lego/blob/e820634e0e43baff0ef61728cb8379f804913b72/providers/dns/cloudflare/cloudflare.go#L34-L35

File: providers/dns/liquidweb/liquidweb.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+// NewDefaultConfig returns a default configuration for the DNSProvider
+func NewDefaultConfig() *Config {
+	config := &Config{
+		URL:                env.GetOrDefaultString("LW_URL", ""),
+		Username:           env.GetOrDefaultString("LW_USERNAME", ""),
+		Password:           env.GetOrDefaultString("LW_PASSWORD", ""),
+		Zone:               env.GetOrDefaultString("LW_ZONE", ""),
+		Timeout:            time.Second * time.Duration(env.GetOrDefaultInt("LW_TIMEOUT", 60)),
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]: Could you follow the same pattern as other provider:
https://github.com/go-acme/lego/blob/e820634e0e43baff0ef61728cb8379f804913b72/providers/dns/linodev4/linodev4.go#L38

File: providers/dns/liquidweb/liquidweb.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	PropagationTimeout time.Duration
+}
+
+// NewDefaultConfig returns a default configuration for the DNSProvider
+func NewDefaultConfig() *Config {
+	config := &Config{
+		URL:                env.GetOrDefaultString("LW_URL", ""),
+		Username:           env.GetOrDefaultString("LW_USERNAME", ""),
+		Password:           env.GetOrDefaultString("LW_PASSWORD", ""),
+		Zone:               env.GetOrDefaultString("LW_ZONE", ""),
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]: URL, Username, Password, zone musr be managed like other provider:

https://github.com/go-acme/lego/blob/e820634e0e43baff0ef61728cb8379f804913b72/providers/dns/cloudflare/cloudflare.go#L51-L63

File: providers/dns/liquidweb/liquidweb.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	dnsEntry, err := d.client.NetworkDNS.Create(params)
+	if err != nil {
+		return fmt.Errorf("could not create TXT record: %v", err)
+	}
+
+	d.recordIDsMu.Lock()
+	d.recordIDs[token] = int(dnsEntry.ID)
+	d.recordIDsMu.Unlock()
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]: could you explains why you are using a lock?

File: providers/dns/liquidweb/liquidweb.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+func NewDNSProvider() (*DNSProvider, error) {
+	config := NewDefaultConfig()
+
+	return NewDNSProviderConfig(config)
+}
+
+// NewDNSProviderConfig return a DNSProvider instance configured for Liquid Web.
+func NewDNSProviderConfig(config *Config) (*DNSProvider, error) {
+	if config == nil {
+		return nil, errors.New("the configuration of the DNS provider is nil")
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]:
```suggestion
		return nil, errors.New("liquidweb: the configuration of the DNS provider is nil")
```

File: providers/dns/liquidweb/liquidweb.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+	if config.Username == "" {
+		return nil, fmt.Errorf("liquidweb: username is missing")
+	}
+
+	if config.Password == "" {
+		return nil, fmt.Errorf("liquidweb: password is missing")
+	}
+
+	// Initial new LW go client.
+	lwAPI, err := lwclient.NewAPI(config.Username, config.Password, config.URL, int(config.Timeout.Seconds()))
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]: Could you rename `lwAPI` to client.

File: providers/dns/liquidweb/liquidweb.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+	"errors"
+	"fmt"
+	"log"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/go-acme/lego/v3/challenge/dns01"
+	"github.com/go-acme/lego/v3/platform/config/env"
+	lwclient "github.com/liquidweb/liquidweb-go/client"
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]: could you rename `lwclient` to `lw`

File: providers/dns/liquidweb/liquidweb.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+	if config.Password == "" {
+		return nil, fmt.Errorf("liquidweb: password is missing")
+	}
+
+	// Initial new LW go client.
+	lwAPI, err := lwclient.NewAPI(config.Username, config.Password, config.URL, int(config.Timeout.Seconds()))
+
+	if err != nil {
+		log.Fatalf("Could not create Liquid Web API client: %v", err)
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]:
```suggestion
		return nil, fmt.Errorf("liquidweb: could not create Liquid Web API client: %v", err)
```

File: providers/dns/liquidweb/liquidweb.go [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+		config:    config,
+		recordIDs: make(map[string]int),
+		client:    lwAPI,
+	}, nil
+}
+
+// Timeout returns the timeout and interval to use when checking for DNS propagation.
+// Adjusting here to cope with spikes in propagation times.
+func (d *DNSProvider) Timeout() (timeout, interval time.Duration) {
+	return d.config.Timeout, d.config.PropagationTimeout
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]:
```suggestion
	return d.config.PropagationTimeout, d.config.PollingInterval
```

File: providers/dns/liquidweb/liquidweb.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+	params := &network.DNSRecordParams{
+		Name:  fqdn[0 : len(fqdn)-1],
+		RData: strconv.Quote(value),
+		Type:  "TXT",
+		Zone:  d.config.Zone,
+	}
+
+	dnsEntry, err := d.client.NetworkDNS.Create(params)
+	if err != nil {
+		return fmt.Errorf("could not create TXT record: %v", err)
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]:
```suggestion
		return fmt.Errorf("liquidweb: could not create TXT record: %v", err)
```

File: providers/dns/liquidweb/liquidweb.go [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+// CleanUp removes the TXT record matching the specified parameters
+func (d *DNSProvider) CleanUp(domain, token, keyAuth string) error {
+	// get the record's unique ID from when we created it
+	d.recordIDsMu.Lock()
+	recordID, ok := d.recordIDs[token]
+	d.recordIDsMu.Unlock()
+	if !ok {
+		return fmt.Errorf("unknown record ID for '%s'", domain)
+	}
+	fmt.Printf("%+v", d.recordIDs)
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]:
```suggestion
```

File: providers/dns/liquidweb/liquidweb.go [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+
+	params := &network.DNSRecordParams{ID: recordID}
+	_, err := d.client.NetworkDNS.Delete(params)
+	if err != nil {
+		return fmt.Errorf("could not remove TXT record: %v", err)
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]:
```suggestion
		return fmt.Errorf("liquidweb: could not remove TXT record: %v", err)
```

File: providers/dns/liquidweb/liquidweb.go [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+// CleanUp removes the TXT record matching the specified parameters
+func (d *DNSProvider) CleanUp(domain, token, keyAuth string) error {
+	// get the record's unique ID from when we created it
+	d.recordIDsMu.Lock()
+	recordID, ok := d.recordIDs[token]
+	d.recordIDsMu.Unlock()
+	if !ok {
+		return fmt.Errorf("unknown record ID for '%s'", domain)
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]:
```suggestion
		return fmt.Errorf(liquidweb: "unknown record ID for %s", domain)
```

File: providers/dns/liquidweb/liquidweb_test.go [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+func TestDNSProvider_Present(t *testing.T) {
+	provider, mux, tearDown := setupTest()
+	defer tearDown()
+
+	mux.HandleFunc("/v1/Network/DNS/Record/create", func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, http.MethodPost, r.Method, "method")
+
+		//assert.Equal(t, "application/json", r.Header.Get("Content-Type"), "Content-Type")
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb_test.go]:
```suggestion
```

File: providers/dns/liquidweb/liquidweb_test.go [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+		assert.Equal(t, "Basic YmxhcnM6dGFjb21hbg==", r.Header.Get("Authorization"), "Authorization")
+
+		reqBody, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb_test.go]:
```suggestion
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
```

File: providers/dns/liquidweb/liquidweb.go (hunk 9 above)
Comment (caiges) [reply providers/dns/liquidweb/liquidweb.go]: Good catch, meant to change this earlier.

File: providers/dns/liquidweb/liquidweb.go (hunk 3 above)
Comment (ldez) [reply providers/dns/liquidweb/liquidweb.go]: `Timeout` -> `HTTPTimeout`
`LIQUID_WEB_TIMEOUT` -> `LIQUID_WEB_HTTP_TIMEOUT`

File: providers/dns/liquidweb/liquidweb.go [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+// NewDefaultConfig returns a default configuration for the DNSProvider
+func NewDefaultConfig() *Config {
+	config := &Config{
+		Timeout:         env.GetOrDefaultSecond("LIQUID_WEB_TIMEOUT", 1*time.Minute),
+		PollingInterval: env.GetOrDefaultSecond("LIQUID_WEB_POLLING_INTERVAL", 2*time.Second),
+		PropagationTimeout: env.GetOrDefaultSecond("LIQUID_WEB_PROPAGATION_TIMEOUT",
+
+			2*time.Minute),
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]: could you remove extra lines and blanks.

File: providers/dns/liquidweb/liquidweb.go [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+	client      *lw.API
+}
+
+// NewDNSProvider returns a DNSProvider instance configured for Liquid Web.
+func NewDNSProvider() (*DNSProvider, error) {
+	config := NewDefaultConfig()
+	config.URL = env.GetOrDefaultString("LIQUID_WEB_URL", "")
+	config.Username = env.GetOrDefaultString("LIQUID_WEB_USERNAME", "")
+	config.Password = env.GetOrDefaultString("LIQUID_WEB_PASSWORD", "")
+	config.Zone = env.GetOrDefaultString("LIQUID_WEB_ZONE", "")
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]: Could you follow this pattern:
https://github.com/go-acme/lego/blob/e820634e0e43baff0ef61728cb8379f804913b72/providers/dns/auroradns/auroradns.go#L47-L59

File: providers/dns/liquidweb/liquidweb.go [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+	Zone               string
+	Timeout            time.Duration
+	PollingInterval    time.Duration
+	PropagationTimeout time.Duration
+}
+
+// NewDefaultConfig returns a default configuration for the DNSProvider
+func NewDefaultConfig() *Config {
+	config := &Config{
+		Timeout:         env.GetOrDefaultSecond("LIQUID_WEB_TIMEOUT", 1*time.Minute),
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]: `Timeout` -> `HTTPTimeout`
`LIQUID_WEB_TIMEOUT` -> `LIQUID_WEB_HTTP_TIMEOUT`

File: providers/dns/liquidweb/liquidweb.toml [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+
+[Configuration]
+  [Configuration.Credentials]
+    LIQUID_WEB_USERNAME = "Storm API Username"
+    LIQUID_WEB_PASSWORD = "Storm API Password"
+  [Configuration.Additional]
+    LIQUID_WEB_URL = "Storm API endpoint"
+    LIQUID_WEB_ZONE = "DNS Zone"
+    LIQUID_WEB_TIMEOUT = "Maximum waiting time for the DNS records to be created (not verified)"
+    LIQUID_WEB_PROPAGATION_TIMEOUT = "Maximum waiting time for DNS propagation"
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.toml]: Cloud you update this file.

File: providers/dns/liquidweb/liquidweb.go [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+		return nil, fmt.Errorf("liquidweb: username is missing")
+	}
+
+	if config.Password == "" {
+		return nil, fmt.Errorf("liquidweb: password is missing")
+	}
+
+	// Initial new LW go client.
+	client, err := lw.NewAPI(config.Username, config.Password, config.URL, int(config.Timeout.Seconds()))
+
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]: could you remove this empty line.

File: providers/dns/liquidweb/liquidweb.go [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+// Present creates a TXT record using the specified parameters
+func (d *DNSProvider) Present(domain, token, keyAuth string) error {
+	fqdn, value := dns01.GetRecord(domain, keyAuth)
+	params := &network.DNSRecordParams{
+		Name:  fqdn[0 : len(fqdn)-1],
+		RData: strconv.Quote(value),
+		Type:  "TXT",
+		Zone:  d.config.Zone,
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]: could you add a configurable TTL:
https://github.com/go-acme/lego/blob/e820634e0e43baff0ef61728cb8379f804913b72/providers/dns/auroradns/auroradns.go#L30

File: providers/dns/liquidweb/liquidweb.go [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+// Adjusting here to cope with spikes in propagation times.
+func (d *DNSProvider) Timeout() (time.Duration, time.Duration) {
+	return d.config.PropagationTimeout, d.config.PollingInterval
+}
+
+// Present creates a TXT record using the specified parameters
+func (d *DNSProvider) Present(domain, token, keyAuth string) error {
+	fqdn, value := dns01.GetRecord(domain, keyAuth)
+	params := &network.DNSRecordParams{
+		Name:  fqdn[0 : len(fqdn)-1],
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]:
```suggestion
		Name:  dns01.UnFqdn(fqdn),
```

File: providers/dns/liquidweb/liquidweb.go (hunk 5 above)
Comment (caiges) [reply providers/dns/liquidweb/liquidweb.go]: I started this work off the back of the Digital Ocean DNS provider. I went through the code and couldn't find anything indicating locks would be necessary and have removed them. Nice catch.

File: providers/dns/liquidweb/liquidweb.go [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	return config
+}
+
+// DNSProvider is an implementation of the acme.ChallengeProvider interface
+// that uses Liquid Web's REST API to manage TXT records for a domain.
+type DNSProvider struct {
+	config    *Config
+	recordIDs map[string]int
```
Comment (ldez) [inline providers/dns/liquidweb/liquidweb.go]: Are you really sure you need that `map`?

**Comment by caiges** [comment]
```
2019/08/18 16:21:48 No key found for account blars@tacoman.com. Generating a P384 key.
2019/08/18 16:21:48 Saved key to /home/caiges/Projects/lego/.lego/accounts/acme-staging-v02.api.letsencrypt.org/blars@tacoman.com/keys/blars@tacoman.com.key
2019/08/18 16:21:48 Please review the TOS at https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf
Do you accept the TOS? Y/n
y
2019/08/18 16:21:50 [INFO] acme: Registering account for blars@tacoman.com
!!!! HEADS UP !!!!

                Your account credentials have been saved in your Let's Encrypt
                configuration directory at "/home/caiges/Projects/lego/.lego/accounts".
                You should make a secure backup of this folder now. This
                configuration directory will also contain certificates and
                private keys obtained from Let's Encrypt so making regular
                backups of this folder is ideal.2019/08/18 16:21:50 [INFO] [*.blars.lwmasre.net, lwmasre.net] acme: Obtaining bundled SAN certificate
2019/08/18 16:21:51 [INFO] [*.blars.lwmasre.net] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/4746302
2019/08/18 16:21:51 [INFO] [lwmasre.net] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/4746303
2019/08/18 16:21:51 [INFO] [*.blars.lwmasre.net] acme: use dns-01 solver
2019/08/18 16:21:51 [INFO] [lwmasre.net] acme: Could not find solver for: tls-alpn-01
2019/08/18 16:21:51 [INFO] [lwmasre.net] acme: Could not find solver for: http-01
2019/08/18 16:21:51 [INFO] [lwmasre.net] acme: use dns-01 solver
2019/08/18 16:21:51 [INFO] [*.blars.lwmasre.net] acme: Preparing to solve DNS-01
2019/08/18 16:21:52 [INFO] [lwmasre.net] acme: Preparing to solve DNS-01
2019/08/18 16:21:54 [INFO] [*.blars.lwmasre.net] acme: Trying to solve DNS-01
2019/08/18 16:21:54 [INFO] [*.blars.lwmasre.net] acme: Checking DNS record propagation using [10.30.7.126:53 10.20.7.126:53 192.168.88.1:53 1.1.1.1:53 8.8.8.8:53 8.8.4.4:53]
2019/08/18 16:21:54 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2019/08/18 16:22:01 [INFO] [*.blars.lwmasre.net] The server validated our request
2019/08/18 16:22:01 [INFO] [lwmasre.net] acme: Trying to solve DNS-01
2019/08/18 16:22:01 [INFO] [lwmasre.net] acme: Checking DNS record propagation using [10.30.7.126:53 10.20.7.126:53 192.168.88.1:53 1.1.1.1:53 8.8.8.8:53 8.8.4.4:53]
2019/08/18 16:22:01 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2019/08/18 16:22:10 [INFO] [lwmasre.net] The server validated our request
2019/08/18 16:22:10 [INFO] [*.blars.lwmasre.net] acme: Cleaning DNS-01 challenge
2019/08/18 16:22:11 [INFO] [lwmasre.net] acme: Cleaning DNS-01 challenge
2019/08/18 16:22:12 [INFO] [*.blars.lwmasre.net, lwmasre.net] acme: Validations succeeded; requesting certificates
2019/08/18 16:22:13 [INFO] [*.blars.lwmasre.net] Server responded with a certificate.
```

File: providers/dns/liquidweb/liquidweb.go (hunk 24 above)
Comment (caiges) [reply providers/dns/liquidweb/liquidweb.go]: The API only accepts an integer ID of the record to delete so it has to be tracked someplace.

Comment (ldez) [reply providers/dns/liquidweb/liquidweb.go]: so you have to add a lock.

**Comment by caiges** [comment]
Here's another run after those changes:

```
2019/08/18 16:59:33 [INFO] [*.tacoblars.lwmasre.net, lwmasre.net] acme: Obtaining bundled SAN certificate
2019/08/18 16:59:34 [INFO] [*.tacoblars.lwmasre.net] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/4746303
2019/08/18 16:59:34 [INFO] [lwmasre.net] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/4750714
2019/08/18 16:59:34 [INFO] [lwmasre.net] acme: authorization already valid; skipping challenge
2019/08/18 16:59:34 [INFO] [*.tacoblars.lwmasre.net] acme: use dns-01 solver
2019/08/18 16:59:34 [INFO] [*.tacoblars.lwmasre.net] acme: Preparing to solve DNS-01
2019/08/18 16:59:36 [INFO] [*.tacoblars.lwmasre.net] acme: Trying to solve DNS-01
2019/08/18 16:59:36 [INFO] [*.tacoblars.lwmasre.net] acme: Checking DNS record propagation using [10.30.7.126:53 10.20.7.126:53 192.168.88.1:53 1.1.1.1:53 8.8.8.8:53 8.8.4.4:53]
2019/08/18 16:59:36 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2019/08/18 16:59:43 [INFO] [*.tacoblars.lwmasre.net] The server validated our request
2019/08/18 16:59:43 [INFO] [*.tacoblars.lwmasre.net] acme: Cleaning DNS-01 challenge
2019/08/18 16:59:44 [INFO] [*.tacoblars.lwmasre.net, lwmasre.net] acme: Validations succeeded; requesting certificates
2019/08/18 16:59:45 [INFO] [*.tacoblars.lwmasre.net] Server responded with a certificate.
```

File: providers/dns/liquidweb/liquidweb.go (hunk 24 above)
Comment (ldez) [reply providers/dns/liquidweb/liquidweb.go]: or use `List` or something else.

Comment (caiges) [reply providers/dns/liquidweb/liquidweb.go]: I might not be looking in the right place but it couldn't see why a lock would be necessary. It looks like the DNS solving mechanism submits them all at once using a standard loop.

Comment (caiges) [reply providers/dns/liquidweb/liquidweb.go]: Same for cleanup

Comment (ldez) [reply providers/dns/liquidweb/liquidweb.go]: a map is not thread safe, lego can be used as a lib, so you need to add a lock if someone by example uses lego inside goroutines .

Comment (caiges) [reply providers/dns/liquidweb/liquidweb.go]: Forgot about the lib use case, updating.

File: go.mod [hunk 25]
```diff
@@ -22,12 +22,14 @@ require (
 	github.com/exoscale/egoscale v0.18.1
 	github.com/go-errors/errors v1.0.1 // indirect
 	github.com/gophercloud/gophercloud v0.3.0
+	github.com/grpc-ecosystem/grpc-gateway v1.9.0 // indirect
```
Comment (ldez) [inline go.mod]: could you remove this line.

File: go.mod [hunk 26]
```diff
@@ -42,10 +44,12 @@ require (
 	github.com/transip/gotransip v0.0.0-20190812104329-6d8d9179b66f
 	github.com/urfave/cli v1.21.0
 	github.com/vultr/govultr v0.1.4
+	go.uber.org/atomic v1.4.0 // indirect
```
Comment (ldez) [inline go.mod]: could you remove this line.

File: go.mod [hunk 27]
```diff
@@ -42,10 +44,12 @@ require (
...
 	golang.org/x/crypto v0.0.0-20190701094942-4def268fd1a4
-	golang.org/x/net v0.0.0-20190503192946-f4e77d36d62c
+	golang.org/x/net v0.0.0-20190522155817-f3200d17e092
 	golang.org/x/oauth2 v0.0.0-20190604053449-0f29369cfe45
 	google.golang.org/api v0.8.0
+	google.golang.org/grpc v1.21.0 // indirect
```
Comment (ldez) [inline go.mod]: could you remove this line.

File: go.mod [hunk 28]
```diff
@@ -42,10 +44,12 @@ require (
 	github.com/transip/gotransip v0.0.0-20190812104329-6d8d9179b66f
 	github.com/urfave/cli v1.21.0
 	github.com/vultr/govultr v0.1.4
+	go.uber.org/atomic v1.4.0 // indirect
 	golang.org/x/crypto v0.0.0-20190701094942-4def268fd1a4
-	golang.org/x/net v0.0.0-20190503192946-f4e77d36d62c
+	golang.org/x/net v0.0.0-20190522155817-f3200d17e092
```
Comment (ldez) [inline go.mod]: could you revert this line.

**Comment by ldez** [comment]
could you do a `go mod tidy`

**Comment by caiges** [comment]
```
LIQUID_WEB_TTL=400 LIQUID_WEB_URL=https://api.stormondemand.com LIQUID_WEB_USERNAME=xxx LIQUID_WEB_PASSWORD=xxx LIQUID_WEB_ZONE=lwmasre.net ./dist/lego -m blars@tacoman.com --dns liquidweb -d "*.tacoblarzii.lwmasre.net" -d lwmasre.net -s https://acme-staging-v02.api.letsencrypt.org/directory run

2019/08/18 17:46:03 No key found for account blars@tacoman.com. Generating a P384 key.
2019/08/18 17:46:03 Saved key to /home/caiges/Projects/lego/.lego/accounts/acme-staging-v02.api.letsencrypt.org/blars@tacoman.com/keys/blars@tacoman.com.key
2019/08/18 17:46:04 Please review the TOS at https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf
Do you accept the TOS? Y/n
y
2019/08/18 17:46:06 [INFO] acme: Registering account for blars@tacoman.com
!!!! HEADS UP !!!!

                Your account credentials have been saved in your Let's Encrypt
                configuration directory at "/home/caiges/Projects/lego/.lego/accounts".
                You should make a secure backup of this folder now. This
                configuration directory will also contain certificates and
                private keys obtained from Let's Encrypt so making regular
                backups of this folder is ideal.2019/08/18 17:46:07 [INFO] [*.tacoblarzii.lwmasre.net, lwmasre.net] acme: Obtaining bundled SAN certificate
2019/08/18 17:46:07 [INFO] [*.tacoblarzii.lwmasre.net] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/4756461
2019/08/18 17:46:07 [INFO] [lwmasre.net] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/4756462
2019/08/18 17:46:07 [INFO] [*.tacoblarzii.lwmasre.net] acme: use dns-01 solver
2019/08/18 17:46:07 [INFO] [lwmasre.net] acme: Could not find solver for: tls-alpn-01
2019/08/18 17:46:07 [INFO] [lwmasre.net] acme: Could not find solver for: http-01
2019/08/18 17:46:07 [INFO] [lwmasre.net] acme: use dns-01 solver
2019/08/18 17:46:07 [INFO] [*.tacoblarzii.lwmasre.net] acme: Preparing to solve DNS-01
2019/08/18 17:46:09 [INFO] [lwmasre.net] acme: Preparing to solve DNS-01
2019/08/18 17:46:11 [INFO] [*.tacoblarzii.lwmasre.net] acme: Trying to solve DNS-01
2019/08/18 17:46:11 [INFO] [*.tacoblarzii.lwmasre.net] acme: Checking DNS record propagation using [10.30.7.126:53 10.20.7.126:53 192.168.88.1:53 1.1.1.1:53 8.8.8.8:53 8.8.4.4:53]
2019/08/18 17:46:11 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2019/08/18 17:46:18 [INFO] [*.tacoblarzii.lwmasre.net] The server validated our request
2019/08/18 17:46:18 [INFO] [lwmasre.net] acme: Trying to solve DNS-01
2019/08/18 17:46:18 [INFO] [lwmasre.net] acme: Checking DNS record propagation using [10.30.7.126:53 10.20.7.126:53 192.168.88.1:53 1.1.1.1:53 8.8.8.8:53 8.8.4.4:53]
2019/08/18 17:46:18 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2019/08/18 17:46:27 [INFO] [lwmasre.net] The server validated our request
2019/08/18 17:46:27 [INFO] [*.tacoblarzii.lwmasre.net] acme: Cleaning DNS-01 challenge
2019/08/18 17:46:28 [INFO] [lwmasre.net] acme: Cleaning DNS-01 challenge
2019/08/18 17:46:29 [INFO] [*.tacoblarzii.lwmasre.net, lwmasre.net] acme: Validations succeeded; requesting certificates
2019/08/18 17:46:30 [INFO] [*.tacoblarzii.lwmasre.net] Server responded with a certificate.
```

**Pull request merged**